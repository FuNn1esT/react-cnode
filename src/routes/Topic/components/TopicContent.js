import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { timeUtils } from 'utils';
import { Card, Button, Avatar, Switch, Icon, Tag, Divider } from 'antd';
import Markdown from 'react-markdown';
import styles from './TopicContent.scss';

const TopicContent = ({
  loading, loadingCollect, topicData,
  renderCollect, renderEdit,
  onClickCollect, onClickEdit,
}) => {
  const author = topicData.author || {};
  const renderTag = (key, isTop, isGood) => {
    const Tabs = {
      top: <Tag color="red">置顶</Tag>,
      good: <Tag color="green">精华</Tag>,
      share: <Tag color="blue">分享</Tag>,
      ask: <Tag color="orange">问答</Tag>,
      job: <Tag color="geekblue">招聘</Tag>,
    };
    if (isTop) {
      return Tabs.top;
    } else if (isGood) {
      return Tabs.good;
    } else {
      return Tabs[key];
    }
  };
  return (
    <Card
      className={styles.container}
      loading={loading}
      extra={renderCollect && (
        <div>
          收藏：
          <Switch
            loading={loadingCollect}
            checked={topicData.is_collect}
            onChange={onClickCollect}
            checkedChildren={<Icon type="check" />}
            unCheckedChildren={<Icon type="cross" />}
          />
        </div>
      )}
      title={(
        <Card.Meta
          avatar={<Avatar
            src={author.avatar_url}
            size="large"
          />}
          title={
            <div>
              {renderTag(topicData.tab, topicData.top, topicData.good)}
              {topicData.title}
            </div>
          }
          description={
            <div className={styles.description}>
              <Link to={`/user/${author.loginname}`}>
                <Tag color="purple">{author.loginname}</Tag>
              </Link>
              <Divider className={styles.divider} type="vertical" />
              <span className={styles.createTime}>
                发布于：{timeUtils.fromNow(topicData.create_at)}
              </span>
              <Divider className={styles.divider} type="vertical" />
              <span className={styles.visitCount}>
                浏览数：{topicData.visit_count}
              </span>
            </div>
          }
        />
      )}
      actions={renderEdit && [
        <Button
          icon="edit"
          onClick={onClickEdit}
        >
          编辑话题
        </Button>,
      ]}
    >
      <Markdown source={topicData.content} />
    </Card>
  );
};

TopicContent.propTypes = {
  loading: PropTypes.bool.isRequired,
  loadingCollect: PropTypes.bool.isRequired,
  topicData: PropTypes.object.isRequired,
  renderCollect: PropTypes.bool.isRequired,
  renderEdit: PropTypes.bool.isRequired,
  onClickCollect: PropTypes.func.isRequired,
  onClickEdit: PropTypes.func.isRequired,
};

export default TopicContent;
