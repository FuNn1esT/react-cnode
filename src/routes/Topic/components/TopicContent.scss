.container {
  width: 1000px;
  max-width: 100%;
  word-break: break-all;
  img {
    max-width: 100%;
  }
}
