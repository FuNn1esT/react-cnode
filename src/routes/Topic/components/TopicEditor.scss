.container {
  background: white;
  margin-top: 24px;
  margin-bottom: 24px;
  max-width: 100%;
  width: 1000px;
}
