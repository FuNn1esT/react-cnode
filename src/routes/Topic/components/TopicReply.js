import React from 'react';
import PropTypes from 'prop-types';
import { List, Avatar } from 'antd';
import Markdown from 'react-markdown';
import styles from './TopicReply.scss';

class TopicReply extends React.PureComponent {
  static propTypes = {
    data: PropTypes.arrayOf(
      PropTypes.shape({
        content: PropTypes.string,
        author: PropTypes.shape({
          avatar_url: PropTypes.string,
          loginname: PropTypes.string,
        }),
      }),
    ),
    loading: PropTypes.bool.isRequired,
  };

  render() {
    const { data, loading } = this.props;
    return (
      <List
        className={styles.container}
        loading={loading}
        dataSource={data}
        renderItem={(item, key) => (
          <List.Item className={styles.content}>
            <List.Item.Meta
              avatar={<Avatar src={item.author.avatar_url} size="large" />}
              title={(
                <div id={item.id}>
                  {item.author.loginname}
                  <a href={`#${item.id}`}>{`${key + 1} æ¥¼`}</a>
                </div>
              )}
              description={
                <Markdown source={item.content} />
              }
            />
          </List.Item>
        )}
      />
    );
  }
}

export default TopicReply;
