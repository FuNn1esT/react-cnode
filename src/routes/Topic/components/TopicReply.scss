.container {
  background: white;
  margin-top: 24px;
  max-width: 100%;
  width: 1000px;
  padding: 0 24px;
}

.content {
  img {
    max-width: 100%;
  }
}
