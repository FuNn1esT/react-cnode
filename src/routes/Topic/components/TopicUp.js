import React from 'react';
import { Rate, Icon } from 'antd';
import styles from './TopicUp.scss';

class TopicUp extends React.PureComponent {
  onUpChange = () => {
    const { id, itemKey } = this.props;
    this.props.onClickUp(id, itemKey);
  };

  render() {
    const { ups, isUp } = this.props;
    const value = isUp === true ? 1 : 0;
    return (
      <div className={styles.container}>
        {ups > 0 ? ups : null}
        <Rate
          className={styles.rate}
          count={1}
          value={value}
          character={<Icon type="like" />}
          onChange={this.onUpChange}
        />
      </div>
    );
  }
}

export default TopicUp;
