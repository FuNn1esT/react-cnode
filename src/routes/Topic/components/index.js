export { default as TopicContent } from './TopicContent';
export { default as TopicReply } from './TopicReply';
export { default as TopicEditor } from './TopicEditor';
