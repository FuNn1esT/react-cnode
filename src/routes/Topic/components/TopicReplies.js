import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { List as ImmutableList } from 'immutable';
import { timeUtils } from 'utils';
import { Card, List, Avatar, Tag, Divider } from 'antd';
import Markdown from 'react-markdown';
import TopicUp from './TopicUp';
import styles from './TopicReplies.scss';

const Replies = ({
  repliesResult,
  current,
  pageSize,
  onReplyPageChange,
  onReplySizeChange,
  onClickUp,
}) => {
  if (repliesResult.size === 0) {
    return null;
  }

  const sliceStart = (current - 1) * pageSize;
  const sliceEnd = sliceStart + pageSize;
  const repliesSlice = repliesResult.slice(sliceStart, sliceEnd);

  return (
    <Card id="topic__reply" className={styles.container} title="回复">
      <List
        dataSource={repliesSlice}
        pagination={{
          total: repliesResult.size,
          current,
          pageSize,
          showSizeChanger: true,
          showQuickJumper: true,
          onChange: onReplyPageChange,
          onShowSizeChange: onReplySizeChange,
        }}
        renderItem={(item, key) => (
          <TopicReply
            replyId={item}
            order={key + 1 + (current - 1) * pageSize}
            onClickUp={onClickUp}
          />
        )}
      />
    </Card>
  );
};

const Reply = ({
  replyId,
  username,
  avatar,
  createAt,
  content,
  order,
  onClickUp,
}) => (
  <List.Item className={styles.content} id={replyId}>
    <List.Item.Meta
      avatar={<Avatar src={avatar} size="large" />}
      title={
        <div className={styles.header}>
          <Link to={`/user/${username}`}>
            <Tag color="purple">{username}</Tag>
          </Link>
          <Divider type="vertical" />
          <a href={`#${replyId}`}>{`${order} 楼`}</a>
          <Divider type="vertical" />
          <span>{timeUtils.fromNow(createAt)}</span>
          <Divider type="vertical" />
          <TopicUp replyId={replyId} onClickUp={onClickUp} />
        </div>
      }
      description={<Markdown className="markdown-body" source={content} />}
    />
  </List.Item>
);

const TopicReplies = connect(state => ({
  repliesResult: state.getIn(['topic', 'repliesData', 'result']),
}))(Replies);

const TopicReply = connect((state, { replyId }) => ({
  username: state.getIn([
    'topic',
    'repliesData',
    'entities',
    replyId,
    'author',
    'loginname',
  ]),
  avatar: state.getIn([
    'topic',
    'repliesData',
    'entities',
    replyId,
    'author',
    'avatar_url',
  ]),
  createAt: state.getIn([
    'topic',
    'repliesData',
    'entities',
    replyId,
    'create_at',
  ]),
  content: state.getIn([
    'topic',
    'repliesData',
    'entities',
    replyId,
    'content',
  ]),
}))(Reply);

Replies.propTypes = {
  repliesResult: PropTypes.instanceOf(ImmutableList).isRequired,
  current: PropTypes.number.isRequired,
  pageSize: PropTypes.number.isRequired,
  onReplyPageChange: PropTypes.func.isRequired,
  onReplySizeChange: PropTypes.func.isRequired,
  onClickUp: PropTypes.func.isRequired,
};

Reply.propTypes = {
  replyId: PropTypes.string.isRequired,
  username: PropTypes.string.isRequired,
  avatar: PropTypes.string.isRequired,
  createAt: PropTypes.string.isRequired,
  content: PropTypes.string.isRequired,
  order: PropTypes.number.isRequired,
  onClickUp: PropTypes.func.isRequired,
};

export default TopicReplies;
