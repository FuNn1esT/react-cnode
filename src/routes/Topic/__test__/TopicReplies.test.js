import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { fromJS, List } from 'immutable';
import toJson from 'enzyme-to-json';
import MockDate from 'mockdate';
import configureStore from 'redux-mock-store';
import { TopicReplies } from '../components';

MockDate.set(new Date('2018-05-21T16:00:00.000Z'));

describe('<TopicReplies />', () => {
  const mockStore = configureStore();
  const initialState = fromJS({
    topic: {
      repliesData: {
        result: ['001'],
        entities: {
          '001': {
            author: {
              loginname: 'admin',
              avatar_url: '//',
            },
            content: 'content',
            create_at: '2018-05-20T16:00:00.000Z',
          },
        },
      },
    },
  });
  const props = {
    current: 1,
    pageSize: 10,
    onReplyPageChange: jest.fn(),
    onReplySizeChange: jest.fn(),
    onClickUp: jest.fn(),
  };

  it('should render correctly', () => {
    const store = mockStore(initialState);
    const tree = mount(
      <Provider store={store}>
        <MemoryRouter>
          <TopicReplies {...props} />
        </MemoryRouter>
      </Provider>,
    );
    expect(toJson(tree, { noKey: true, mode: 'deep' })).toMatchSnapshot();
  });

  it('should render null', () => {
    const store = mockStore(
      initialState.setIn(['topic', 'repliesData', 'result'], List([])),
    );
    const tree = mount(
      <Provider store={store}>
        <MemoryRouter>
          <TopicReplies {...props} />
        </MemoryRouter>
      </Provider>,
    );
    expect(toJson(tree, { noKey: true, mode: 'deep' })).toMatchSnapshot();
  });
});
