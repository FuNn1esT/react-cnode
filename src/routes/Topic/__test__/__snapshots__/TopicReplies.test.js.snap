// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TopicReplies /> should render correctly 1`] = `
<div
  className="ant-card ant-card-bordered"
  id="topic__reply"
>
  <div
    className="ant-card-head"
  >
    <div
      className="ant-card-head-wrapper"
    >
      <div
        className="ant-card-head-title"
      >
        回复
      </div>
    </div>
  </div>
  <div
    className="ant-card-body"
    style={Object {}}
  >
    <div
      className="ant-list ant-list-split ant-list-something-after-last-item"
    >
      <div
        className="ant-spin-nested-loading"
        style={null}
      >
        <div
          className="ant-spin-container"
        >
          <div
            className="ant-list-item"
            id="001"
          >
            <div
              className="ant-list-item-meta"
            >
              <div
                className="ant-list-item-meta-avatar"
              >
                <span
                  className="ant-avatar ant-avatar-lg ant-avatar-circle ant-avatar-image"
                >
                  <img
                    onError={[Function]}
                    src="//"
                  />
                </span>
              </div>
              <div
                className="ant-list-item-meta-content"
              >
                <h4
                  className="ant-list-item-meta-title"
                >
                  <div>
                    <a
                      href="/user/admin"
                      onClick={[Function]}
                    >
                      <div
                        className="ant-tag ant-tag-green"
                        data-show={true}
                        style={
                          Object {
                            "backgroundColor": null,
                          }
                        }
                      >
                        admin
                      </div>
                    </a>
                    <div
                      className="ant-divider ant-divider-vertical"
                    />
                    <a
                      href="#001"
                    >
                      1 楼
                    </a>
                    <div
                      className="ant-divider ant-divider-vertical"
                    />
                    <span>
                      a day ago
                    </span>
                    <div
                      className="ant-divider ant-divider-vertical"
                    />
                    <div>
                      <ul
                        className="ant-rate"
                        onBlur={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onMouseLeave={[Function]}
                        style={Object {}}
                        tabIndex={0}
                      >
                        <li
                          className="ant-rate-star ant-rate-star-zero"
                          onClick={[Function]}
                          onMouseMove={[Function]}
                        >
                          <div
                            className="ant-rate-star-first"
                          >
                            <i
                              className="anticon anticon-like"
                            />
                          </div>
                          <div
                            className="ant-rate-star-second"
                          >
                            <i
                              className="anticon anticon-like"
                            />
                          </div>
                        </li>
                      </ul>
                    </div>
                  </div>
                </h4>
                <div
                  className="ant-list-item-meta-description"
                >
                  <div
                    className="markdown-body"
                  >
                    <p>
                      content
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="ant-list-item"
            id="002"
          >
            <div
              className="ant-list-item-meta"
            >
              <div
                className="ant-list-item-meta-avatar"
              >
                <span
                  className="ant-avatar ant-avatar-lg ant-avatar-circle ant-avatar-image"
                >
                  <img
                    onError={[Function]}
                    src="//"
                  />
                </span>
              </div>
              <div
                className="ant-list-item-meta-content"
              >
                <h4
                  className="ant-list-item-meta-title"
                >
                  <div>
                    <a
                      href="/user/guest"
                      onClick={[Function]}
                    >
                      <div
                        className="ant-tag ant-tag-purple"
                        data-show={true}
                        style={
                          Object {
                            "backgroundColor": null,
                          }
                        }
                      >
                        guest
                      </div>
                    </a>
                    <div
                      className="ant-divider ant-divider-vertical"
                    />
                    <a
                      href="#002"
                    >
                      2 楼
                    </a>
                    <div
                      className="ant-divider ant-divider-vertical"
                    />
                    <span>
                      a day ago
                    </span>
                    <div
                      className="ant-divider ant-divider-vertical"
                    />
                    <div>
                      <ul
                        className="ant-rate"
                        onBlur={[Function]}
                        onFocus={[Function]}
                        onKeyDown={[Function]}
                        onMouseLeave={[Function]}
                        style={Object {}}
                        tabIndex={0}
                      >
                        <li
                          className="ant-rate-star ant-rate-star-zero"
                          onClick={[Function]}
                          onMouseMove={[Function]}
                        >
                          <div
                            className="ant-rate-star-first"
                          >
                            <i
                              className="anticon anticon-like"
                            />
                          </div>
                          <div
                            className="ant-rate-star-second"
                          >
                            <i
                              className="anticon anticon-like"
                            />
                          </div>
                        </li>
                      </ul>
                    </div>
                  </div>
                </h4>
                <div
                  className="ant-list-item-meta-description"
                >
                  <div
                    className="markdown-body"
                  >
                    <p>
                      guest content
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="ant-list-pagination"
      >
        <ul
          className="ant-pagination "
          style={Object {}}
          unselectable="unselectable"
        >
          <li
            aria-disabled={true}
            className="ant-pagination-disabled ant-pagination-prev"
            onClick={[Function]}
            onKeyPress={[Function]}
            tabIndex={null}
            title="Previous Page"
          >
            <a
              className="ant-pagination-item-link"
            />
          </li>
          <li
            className="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
            onClick={[Function]}
            onKeyPress={[Function]}
            tabIndex="0"
            title={1}
          >
            <a>
              1
            </a>
          </li>
          <li
            aria-disabled={true}
            className="ant-pagination-disabled ant-pagination-next"
            onClick={[Function]}
            onKeyPress={[Function]}
            tabIndex={null}
            title="Next Page"
          >
            <a
              className="ant-pagination-item-link"
            />
          </li>
          <li
            className="ant-pagination-options"
          >
            <div
              className="ant-pagination-options-size-changer ant-select ant-select-enabled"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
            >
              <div
                aria-autocomplete="list"
                aria-expanded={false}
                aria-haspopup="true"
                className="ant-select-selection
            ant-select-selection--single"
                onKeyDown={[Function]}
                role="combobox"
                tabIndex={0}
              >
                <div
                  className="ant-select-selection__rendered"
                >
                  <div
                    className="ant-select-selection-selected-value"
                    style={
                      Object {
                        "display": "block",
                        "opacity": 1,
                      }
                    }
                    title="10 / page"
                  >
                    10 / page
                  </div>
                </div>
                <span
                  className="ant-select-arrow"
                  onClick={[Function]}
                  style={
                    Object {
                      "WebkitUserSelect": "none",
                      "userSelect": "none",
                    }
                  }
                  unselectable="on"
                >
                  <b />
                </span>
              </div>
            </div>
            <div
              className="ant-pagination-options-quick-jumper"
            >
              Goto
              <input
                onChange={[Function]}
                onKeyUp={[Function]}
                type="text"
                value=""
              />
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;

exports[`<TopicReplies /> should render null 1`] = `""`;
